src/Model/Command/AbstractUserRecommendation.php

```diff
@@ -0,0 +1,104 @@
+<?php declare(strict_types=1);
+
+namespace Lmc\Matej\Model\Command;
+
+use Lmc\Matej\Model\Assertion;
+
+/**
+ * Deliver recommendations for given user.
+ */
+abstract class AbstractUserRecommendation extends AbstractRecommendation implements UserAwareInterface
+{
+    /** @var string */
+    private $userId;
+    /** @var float */
+    private $rotationRate;
+    /** @var int */
+    private $rotationTime;
+    /** @var bool */
+    private $hardRotation;
+
+    protected function __construct(string $userId, string $scenario)
+    {
+        parent::__construct($scenario);
+        $this->setUserId($userId);
+    }
+
+    /**
+     * Even with rotation rate 1.0 user could still obtain the same recommendations in some edge cases.
+     * To prevent this, enable hard rotation - recommended items are then excluded until rotation time is expired.
+     * By default hard rotation is not enabled.
+     * @return $this
+     */
+    public function enableHardRotation(): self
+    {
+        $this->hardRotation = true;
+
+        return $this;
+    }
+
+    /**
+     * Get id of user requesting recommendation.
+     */
+    public function getUserId(): string
+    {
+        return $this->userId;
+    }
+
+    /**
+     * Set how much should the item be penalized for being recommended again in the near future.
+     *
+     * @return $this
+     */
+    public function setRotationRate(float $rotationRate): self
+    {
+        Assertion::between($rotationRate, 0, 1);
+
+        $this->rotationRate = $rotationRate;
+
+        return $this;
+    }
+
+    /**
+     * Specify for how long will the item's rotationRate be taken in account and so the item is penalized for
+     * recommendations.
+     *
+     * @return $this
+     */
+    public function setRotationTime(int $rotationTime): self
+    {
+        Assertion::greaterOrEqualThan($rotationTime, 0);
+
+        $this->rotationTime = $rotationTime;
+
+        return $this;
+    }
+
+    protected function setUserId(string $userId): void
+    {
+        Assertion::typeIdentifier($userId);
+
+        $this->userId = $userId;
+    }
+
+    protected function getCommandParameters(): array
+    {
+        $parameters = parent::getCommandParameters();
+
+        $parameters['user_id'] = $this->userId;
+
+        if ($this->rotationRate !== null) {
+            $parameters['rotation_rate'] = $this->rotationRate;
+        }
+
+        if ($this->rotationRate !== null) {
```

Comment:
```suggestion
        if ($this->rotationTime !== null) {
```

It also means this is not covered by a test 🤔

Reply:
🔝 This bug is still present (and few other comments as well)

Reply:
The problem is that we don't have a test that would test parameters one by one. It didn't appear in coverage because 1) we have a test when rotationRate is set and 2) when it isn't so technically all branches were covered. But branch coverage is not enough in this case. I added test showing this particular bug was fixed but I don't plan to add a matrix test of all the combinations that are possible. It would be around 16M tests.